src/Splitio-net-core/Services/Impressions/Classes/ImpressionsManager.cs

```diff
@@ -0,0 +1,96 @@
+﻿using Splitio.Domain;
+using Splitio.Services.Impressions.Interfaces;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Splitio.Services.Impressions.Classes
+{
+    public class ImpressionsManager : IImpressionsManager
+    {
+        private readonly IImpressionsObserver _impressionsObserver;
+        private readonly IImpressionsLog _impressionsLog;
+        private readonly IImpressionListener _customerImpressionListener;
+        private readonly IImpressionsCounter _impressionsCounter;
+        private readonly bool _optimized;
+        private readonly bool _addPreviousTime;
+
+        public ImpressionsManager(IImpressionsLog impressionsLog,
+            IImpressionListener customerImpressionListener,
+            IImpressionsCounter impressionsCounter,
+            bool addPreviousTime,
+            ImpressionMode impressionMode,
+            IImpressionsObserver impressionsObserver = null)
+        {            
+            _impressionsLog = impressionsLog;
+            _customerImpressionListener = customerImpressionListener;
+            _impressionsCounter = impressionsCounter;
+            _addPreviousTime = addPreviousTime;
+            _optimized = impressionMode == ImpressionMode.Optimized && addPreviousTime;
+            _impressionsObserver = impressionsObserver;
+        }
+
+        public KeyImpression BuildImpression(string matchingKey, string feature, string treatment, long time, long? changeNumber, string label, string bucketingKey)
+        {
+            var impression = new KeyImpression(matchingKey, feature, treatment, time, changeNumber, label, bucketingKey);
+
+            if (_addPreviousTime && _impressionsObserver != null)
```

Comment:
you could add this validation in the constructor and only evaluate one of this here (false or true)